sylt-compiler/src/typechecker.rs

```diff
@@ -1594,6 +1675,17 @@ impl TypeChecker {
 
                     _ => Ok(()),
                 },
+
+                Constraint::Pure => match self.find_type(a) {
+                    Type::Function(_, _) => Ok(()),
+                    Type::Unknown => Ok(()),
+                    _ => err_type_error!(
+                        self,
+                        span,
+                        TypeError::Impurity,
+                        "Non pure function something something"
+                    ),
+                },
```

Comment:
I am not actually certain what this code does, hence the bad message. Although I will look in to it.

Reply:
Is it that only functions (or unknowns) can have the pure constraint?

Reply:
You end up here if you have something that isn't a function, that has the "Pure" tag on it. So if you managed to put in something that cannot be Pure.

Here's an example that should trigger that error:
```
a : fn <*a: Pure> *a -> *a : fn a -> int do a end
```
So your second comment is correct. A better error message might be:
> Only functions can be pure, but this is not a function
Someone else will hve more opinions on the spelling and formulation. But something along those lines.

Reply:
It's also a good idea to use the `violating` type error if you have a type that you can point at. That way we print the type that is wrong. ;)